File: apps/web/lib/engine/runner.ts
import { serviceClient } from "@/lib/supabase/service";
import { extractPackages } from "./extractor";
import { matchCves } from "./matcher";
import { checkStaleness } from "./staleness";

export async function runScan(scanId: string): Promise<void> {
  // Mark scan as processing
  await serviceClient
    .from("scans")
    .update({ status: "processing" })
    .eq("id", scanId);

  try {
    // Fetch scan files
    const { data: files } = await serviceClient
      .from("scan_files")
      .select("*")
      .eq("scan_id", scanId);

    if (!files || files.length === 0) {
      await serviceClient
        .from("scans")
        .update({ status: "failed" })
        .eq("id", scanId);
      return;
    }

    const allVulnerabilities = [];

    for (const file of files) {
      // Extract packages from code
      const packages = extractPackages(file.content, file.language);

      // Run CVE matching and staleness checking in parallel
      const [cveResults, stalenessResults] = await Promise.all([
        matchCves(packages),
        checkStaleness(file.content, packages),
      ]);

      const combined = [...cveResults, ...stalenessResults];

      for (const vuln of combined) {
        allVulnerabilities.push({
          scan_id: scanId,
          file_id: file.id,
          type: vuln.type,
          severity: vuln.severity,
          line_start: vuln.lineStart,
          line_end: vuln.lineEnd,
          description: vuln.description,
          cve_id: vuln.cveId,
          suggestion: vuln.suggestion,
        });
      }
    }

    // Insert all vulnerabilities
    if (allVulnerabilities.length > 0) {
      await serviceClient
        .from("vulnerabilities")
        .insert(allVulnerabilities);
    }

    // Mark scan as completed
    await serviceClient
      .from("scans")
      .update({
        status: "completed",
        completed_at: new Date().toISOString(),
      })
      .eq("id", scanId);

  } catch (error) {
    console.error("Scan failed:", error);
    await serviceClient
      .from("scans")
      .update({ status: "failed" })
      .eq("id", scanId);
  }
}

File: apps/analysis-engine/internal/runner/runner.go
package runner

import (
	"fmt"
	"log"

	supabase "github.com/supabase-community/supabase-go"
	"github.com/yuricommits/hwa/analysis-engine/internal/extractor"
	"github.com/yuricommits/hwa/analysis-engine/internal/matcher"
	"github.com/yuricommits/hwa/analysis-engine/internal/staleness"
)

type ScanFile struct {
	ID       string `json:"id"`
	ScanID   string `json:"scan_id"`
	Filename string `json:"filename"`
	Content  string `json:"content"`
	Language string `json:"language"`
}

type Vulnerability struct {
	ScanID      string  `json:"scan_id"`
	FileID      string  `json:"file_id"`
	Type        string  `json:"type"`
	Severity    string  `json:"severity"`
	LineStart   *int    `json:"line_start"`
	LineEnd     *int    `json:"line_end"`
	Description string  `json:"description"`
	CveID       *string `json:"cve_id"`
	Suggestion  string  `json:"suggestion"`
}

func RunScan(db *supabase.Client, scanID string) error {
	// Mark as processing
	_, _, err := db.From("scans").
		Update(map[string]any{"status": "processing"}, "", "").
		Eq("id", scanID).
		Execute()
	if err != nil {
		return fmt.Errorf("failed to update scan status: %w", err)
	}

	// Fetch scan files
	var files []ScanFile
	_, err = db.From("scan_files").
		Select("*", "", false).
		Eq("scan_id", scanID).
		ExecuteTo(&files)
	if err != nil {
		markFailed(db, scanID)
		return fmt.Errorf("failed to fetch scan files: %w", err)
	}

	if len(files) == 0 {
		markFailed(db, scanID)
		return fmt.Errorf("no files found for scan %s", scanID)
	}

	var allVulns []Vulnerability

	for _, file := range files {
		// Extract packages
		packages := extractor.ExtractPackages(file.Content, file.Language)

		// Match CVEs
		cveVulns, err := matcher.MatchCves(db, packages)
		if err != nil {
			log.Printf("CVE matching failed for file %s: %v", file.ID, err)
		}

		// Check staleness
		stalenessVulns, err := staleness.CheckStaleness(db, file.Content, packages, file.Language)
		if err != nil {
			log.Printf("Staleness check failed for file %s: %v", file.ID, err)
		}

		for _, v := range append(cveVulns, stalenessVulns...) {
			allVulns = append(allVulns, Vulnerability{
				ScanID:      scanID,
				FileID:      file.ID,
				Type:        v.Type,
				Severity:    v.Severity,
				LineStart:   v.LineStart,
				LineEnd:     v.LineEnd,
				Description: v.Description,
				CveID:       v.CveID,
				Suggestion:  v.Suggestion,
			})
		}
	}

	// Insert vulnerabilities
	if len(allVulns) > 0 {
		_, _, err = db.From("vulnerabilities").
			Insert(allVulns, false, "", "", "").
			Execute()
		if err != nil {
			markFailed(db, scanID)
			return fmt.Errorf("failed to insert vulnerabilities: %w", err)
		}
	}

	// Mark completed
	_, _, err = db.From("scans").
		Update(map[string]any{
			"status":       "completed",
			"completed_at": "now()",
		}, "", "").
		Eq("id", scanID).
		Execute()
	if err != nil {
		return fmt.Errorf("failed to mark scan completed: %w", err)
	}

	log.Printf("Scan %s completed with %d vulnerabilities", scanID, len(allVulns))
	return nil
}

func markFailed(db *supabase.Client, scanID string) {
	db.From("scans").
		Update(map[string]any{"status": "failed"}, "", "").
		Eq("id", scanID).
		Execute()
}
